File: apps/api/app/api/tasks.py
"""My Work — the task inbox.

**These routes carry no permission dependency, and that is deliberate.**
Every other route in this application sits behind `require_permission`.
My Work does not, because it returns only the caller's own tasks and the
unclaimed tasks addressed to roles the caller actually holds. There is no
permission to check: the query is scoped by `principal.user_id` and
`principal.roles`, both of which come from a signature-verified token and
a database lookup, never from the request.

Requiring a permission here would mean a user could be locked out of
their own inbox by an administrator forgetting a grant -- work would pile
up addressed to somebody who cannot see it, and nothing would report the
condition.

The one exception is :func:`create_task`, which addresses work to
*other* people and therefore needs `project.edit`.
"""

from __future__ import annotations

import uuid
from datetime import date
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from app.core.security import (
    Principal,
    get_db,
    get_principal,
    require_permission,
    require_project_member,
)
from app.core.tenancy import CrossTenantReferenceError
from app.domains.tasks.service import (
    TaskInput,
    TaskNotFoundError,
    TaskStateError,
    claim_task,
    complete_task,
    create_task,
    my_work,
    my_work_counts,
    project_tasks,
    reassign_task,
)

router = APIRouter()

__all__ = ["router"]


class TaskCreate(BaseModel):
    task_type: str = Field(min_length=1, max_length=50)
    title: str = Field(min_length=1, max_length=200)
    description: str | None = None
    priority: str = Field(default="medium", pattern="^(low|medium|high|critical)$")
    assigned_user_id: uuid.UUID | None = None
    assigned_role: str | None = None
    project_id: uuid.UUID | None = None
    due_date: date | None = None
    required_action: str | None = None
    entity_type: str | None = None
    entity_id: uuid.UUID | None = None


class TaskComplete(BaseModel):
    outcome_note: str | None = None


class TaskReassign(BaseModel):
    to_user_id: uuid.UUID
    # Required, not optional. "Why is this suddenly mine" is the first
    # question the new assignee asks.
    reason: str = Field(min_length=3, max_length=500)


def _refuse(exc: Exception) -> HTTPException:
    return HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))


def _bad_reference(exc: CrossTenantReferenceError) -> HTTPException:
    """400, not 403 or 404.

    The caller named a user who is not a member of this organization. 403
    would imply the id is real and merely off-limits; 404 would imply it
    is not real. Both leak. 400 says only that the payload was wrong,
    which is all the caller is entitled to know.
    """
    return HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc))


@router.get("", tags=["my-work"])
def list_my_work(
    include_done: bool = Query(default=False),
    principal: Principal = Depends(get_principal),
    session: Session = Depends(get_db),
) -> list[dict[str, Any]]:
    """The caller's inbox, most urgent first."""
    return my_work(
        session,
        user_id=principal.user_id,
        organization_id=principal.organization_id,
        role_codes=principal.roles,
        include_done=include_done,
    )


@router.get("/counts", tags=["my-work"])
def get_my_work_counts(
    principal: Principal = Depends(get_principal),
    session: Session = Depends(get_db),
) -> dict[str, int]:
    """Sidebar badge numbers.

    Actionable items only (CLAUDE.md §11). Built from the same predicate
    as the list above, so the badge and the inbox can never disagree.
    """
    return my_work_counts(
        session,
        user_id=principal.user_id,
        organization_id=principal.organization_id,
        role_codes=principal.roles,
    )


@router.post("", status_code=status.HTTP_201_CREATED, tags=["my-work"])
def post_task(
    payload: TaskCreate,
    principal: Principal = Depends(require_permission("project.edit")),
    session: Session = Depends(get_db),
) -> dict[str, str]:
    """Raise a task. Addresses work to another person or role."""
    try:
        task_id = create_task(
            session,
            data=TaskInput(
                task_type=payload.task_type,
                title=payload.title,
                description=payload.description,
                priority=payload.priority,
                assigned_user_id=payload.assigned_user_id,
                assigned_role=payload.assigned_role,
                project_id=payload.project_id,
                due_date=payload.due_date,
                source_event="manual",
                entity_type=payload.entity_type,
                entity_id=payload.entity_id,
                required_action=payload.required_action,
            ),
            actor_id=principal.user_id,
            organization_id=principal.organization_id,
        )
    except CrossTenantReferenceError as exc:
        raise _bad_reference(exc) from exc
    except TaskStateError as exc:
        raise _refuse(exc) from exc
    return {"id": str(task_id)}


@router.post("/{task_id}/claim", status_code=status.HTTP_204_NO_CONTENT, tags=["my-work"])
def post_claim(
    task_id: uuid.UUID,
    principal: Principal = Depends(get_principal),
    session: Session = Depends(get_db),
) -> None:
    """Take a role-addressed task.

    No permission check: the service refuses any task not addressed to a
    role the caller holds, which is a stronger condition than any
    permission grant could express.
    """
    try:
        claim_task(
            session,
            task_id=task_id,
            user_id=principal.user_id,
            organization_id=principal.organization_id,
            role_codes=principal.roles,
        )
    except TaskStateError as exc:
        raise _refuse(exc) from exc


@router.post("/{task_id}/complete", status_code=status.HTTP_204_NO_CONTENT, tags=["my-work"])
def post_complete(
    task_id: uuid.UUID,
    payload: TaskComplete,
    principal: Principal = Depends(get_principal),
    session: Session = Depends(get_db),
) -> None:
    """Mark the caller's own task done. The service refuses anyone else's."""
    try:
        complete_task(
            session,
            task_id=task_id,
            actor_id=principal.user_id,
            organization_id=principal.organization_id,
            outcome_note=payload.outcome_note,
        )
    except TaskNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except TaskStateError as exc:
        raise _refuse(exc) from exc


@router.post("/{task_id}/reassign", status_code=status.HTTP_204_NO_CONTENT, tags=["my-work"])
def post_reassign(
    task_id: uuid.UUID,
    payload: TaskReassign,
    principal: Principal = Depends(require_permission("project.edit")),
    session: Session = Depends(get_db),
) -> None:
    """Delegate a task. Requires the authority to direct someone else's work."""
    try:
        reassign_task(
            session,
            task_id=task_id,
            to_user_id=payload.to_user_id,
            actor_id=principal.user_id,
            organization_id=principal.organization_id,
            reason=payload.reason,
        )
    except CrossTenantReferenceError as exc:
        raise _bad_reference(exc) from exc
    except TaskNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except TaskStateError as exc:
        raise _refuse(exc) from exc


@router.get("/project/{project_id}", tags=["my-work"])
def list_project_tasks(
    project_id: uuid.UUID,
    principal: Principal = Depends(require_permission("project.view")),
    _scope: Principal = Depends(require_project_member("project_id")),
    session: Session = Depends(get_db),
) -> list[dict[str, Any]]:
    """Everything outstanding on a project, not just the caller's share.

    Both dependencies, as everywhere: `project.view` asks whether this
    person may ever read a project, `require_project_member` asks whether
    they may read *this* one.
    """
    return project_tasks(session, project_id=project_id, organization_id=principal.organization_id)

File: apps/web/lib/api/tasks.ts
/**
 * My Work, over HTTP.
 *
 * 🔴 THIS LIST IS AN INBOX, NOT A TABLE OF TASKS.
 *
 * `my_work` returns the caller's own tasks PLUS unclaimed tasks addressed
 * to their roles, ordered overdue-first. Two consequences the UI must not
 * flatten:
 *
 *   * `assigned_user_id` is null on a role-addressed item. That is not
 *     missing data — it is the definition of "nobody has claimed this
 *     yet", and it is why the item is in this list at all.
 *   * `is_overdue` is computed by the SERVER against `CURRENT_DATE`.
 *     The browser must not re-derive it from `due_date`: the two clocks
 *     disagree across time zones, and a task that the server calls
 *     overdue while the screen calls it due today is a contradiction the
 *     reader cannot resolve.
 *
 * `CLAUDE.md` §11 also requires that counts here be ACTIONABLE items
 * rather than totals — which is what this endpoint already returns, since
 * it filters to actionable statuses unless `include_done` is set.
 */

import { z } from "zod";

import { apiRequest, type ApiCredentials } from "./client";

export const taskSchema = z.object({
  id: z.string(),
  task_type: z.string(),
  title: z.string(),
  description: z.string().nullable(),
  priority: z.string(),
  status: z.string(),
  due_date: z.string().nullable(),
  required_action: z.string().nullable(),
  entity_type: z.string().nullable(),
  entity_id: z.string().nullable(),
  project_id: z.string().nullable(),
  assigned_user_id: z.string().nullable(),
  assigned_role: z.string().nullable(),
  // NOT NULL in `workflow.tasks`. See the formulations schema for why a
  // needlessly nullable field hides a contract regression.
  created_at: z.string(),
  // From the LEFT JOIN: null for a task that belongs to no project.
  project_code: z.string().nullable(),
  project_name: z.string().nullable(),
  // Server-computed. See the header: do NOT re-derive this in the browser.
  is_overdue: z.boolean(),
});

export type Task = z.infer<typeof taskSchema>;

const taskList = z.array(taskSchema);

export function fetchMyWork(
  credentials: ApiCredentials,
  signal?: AbortSignal,
): Promise<Task[]> {
  return apiRequest({ path: "/api/my-work", credentials, signal }, (payload) =>
    taskList.parse(payload),
  );
}

/*
 * 🔴 `/api/my-work`, NOT `/api/tasks`. THE ROUTER IS MOUNTED UNDER THE QUEUE.
 *
 * `main.py` does `include_router(tasks_router, prefix="/api/my-work")`, so
 * every task route -- create, claim, complete, reassign -- lives under the
 * name of the SCREEN rather than of the table. `createTask` was written
 * against `/api/tasks` and 404'd on every press. The live suite found it: a
 * 404 from a path that does not exist looks exactly like a refusal from one
 * that does, and nothing below the browser was wrong.
 *
 * ⚠️ AND THE PREFIX IS WRITTEN OUT IN FULL AT EACH CALL, NOT HOISTED INTO A
 * CONSTANT. Hoisting it was the obvious tidy-up and it silently disabled the
 * guard that catches this: `every path the web client calls is a path the API
 * serves` reads these literals out of the source, and `path: TASKS` and
 * `` `${TASKS}/${id}/claim` `` are invisible to it -- measured, by putting the
 * wrong base back and watching the check report nothing missing. A literal
 * that a reader and a test can both see beats a constant neither can.
 */
export interface TaskCreateRequest {
  readonly task_type: string;
  readonly title: string;
  readonly description?: string;
  readonly priority?: string;
  /**
   * 🔴 ONE OF THESE TWO IS REQUIRED, AND THE SERVER REFUSES A TASK WITHOUT ONE.
   *
   * `create_task` raises *"a task needs an owner: give it an assigned user or
   * an assigned role"*, and a CHECK constraint backs it. They were absent from
   * this type, so the Raise-a-task form sent neither and every press returned
   * 409 — the form had never once created a task. `my-work` selects on both
   * (`assigned_user_id = :uid OR (assigned_user_id IS NULL AND assigned_role =
   * ANY(:roles))`), so a role-addressed task reaches everyone holding it until
   * somebody takes it.
   *
   * Not modelled as a union: the server takes both fields and this mirrors the
   * server. The FORM is what enforces "exactly one", because that is a question
   * about the control, not about the wire.
   */
  readonly assigned_user_id?: string;
  readonly assigned_role?: string;
  readonly project_id?: string;
  readonly due_date?: string;
  readonly required_action?: string;
}

export const TASK_PRIORITIES = ["low", "medium", "high", "critical"] as const;

/**
 * The ten seeded realm roles, for addressing a task to a role rather than a
 * person.
 *
 * ⚠️ A MIRROR OF `apps/api/migrations/002_seed_roles_permissions.sql`, and
 * `tasks.roles.drift.test.ts` reads that file to prove it still matches. Two
 * literals in two files cannot be type-checked into agreement.
 *
 * 🔴 THE REASON THIS LIST CAN EXIST AT ALL is that a role is not a person.
 * `RaiseTaskForm` recorded that it had no assignee field because no endpoint
 * lists the organization's people for a non-administrator — true, and it never
 * applied to roles. The catalogue is fixed, seeded, and not confidential;
 * `GET /api/admin/roles` needs `admin.roles` only because it also reports each
 * role's GRANTS, which is a different question from what the roles are called.
 */
export const ASSIGNABLE_ROLES = [
  { code: "product_development_chemist", label: "Product Development Chemist" },
  { code: "product_development_engineer", label: "Product Development Engineer" },
  { code: "product_development_lead", label: "Product Development Lead" },
  { code: "product_development_director", label: "Product Development Director" },
  { code: "qa_compliance_officer", label: "QA / Compliance Officer" },
  { code: "laboratory_technician", label: "Laboratory Technician" },
  { code: "procurement_specialist", label: "Procurement / Material Specialist" },
  { code: "production_engineer", label: "Production Engineer" },
  { code: "executive_viewer", label: "Executive Viewer" },
  { code: "administrator", label: "Administrator" },
] as const;

export function createTask(
  credentials: ApiCredentials,
  request: TaskCreateRequest,
): Promise<{ id: string }> {
  return apiRequest(
    { path: "/api/my-work", method: "POST", credentials, body: request },
    (payload) => z.object({ id: z.string() }).passthrough().parse(payload),
  );
}
/**
 * Take an unclaimed, role-addressed task for yourself.
 *
 * 🔴 THIS ROUTE, `/complete` AND `/reassign` HAD NO CLIENT AT ALL. A task could
 * be raised and could never be picked up, finished or handed on through the
 * browser — the whole second half of the task lifecycle was unreachable, and
 * `/reassign` IS "assign this task to that person".
 *
 * No permission on the route on purpose: claiming is scoped by what the queue
 * already shows you, and `my_work` filters that by your roles.
 */
export function claimTask(credentials: ApiCredentials, taskId: string): Promise<unknown> {
  return apiRequest(
    { path: `/api/my-work/${taskId}/claim`, method: "POST", credentials, body: {} },
    (payload) => payload,
  );
}

/** Mark a task done, optionally saying what came of it. */
export function completeTask(
  credentials: ApiCredentials,
  taskId: string,
  outcomeNote?: string,
): Promise<unknown> {
  return apiRequest(
    {
      path: `/api/my-work/${taskId}/complete`,
      method: "POST",
      credentials,
      body: outcomeNote === undefined ? {} : { outcome_note: outcomeNote },
    },
    (payload) => payload,
  );
}

/*
 * ⚠️ `POST /api/my-work/{id}/reassign` HAS NO CLIENT HERE, ON PURPOSE.
 *
 * It is the third orphaned route on this module and the one that most deserves
 * a control -- handing work to a named colleague. It is not built yet because
 * it needs a PEOPLE source and there is exactly one a non-administrator can
 * read: `GET /api/projects/{id}/members`. So the control belongs beside a task
 * that HAS a project, and `TaskRow` does not carry `project_id` today.
 *
 * Writing the client now and the control later is precisely the defect this
 * module's own history records -- a request function is not a caller, and an
 * uncalled one reads like a feature that exists. It ships with its form.
 */
